File: turtlesim_catch_all_turtles/turtlesim_catch_all_turtles/turtle_spawner.py
#!/usr/bin/python3

import rclpy
from rclpy.node import Node
from turtlesim.srv import Spawn
from turtlesim.srv import Kill
from random import randint
from random import random
from functools import partial
from math import pi
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.msg import Turtles
from my_robot_interfaces.srv import KillTurtle


class turtleSpawnerNode(Node):
    def __init__(self):
        super().__init__("turtle_spawner")

        # Parameters
        self.declare_parameter("spawn_frequency", 1)
        self.declare_parameter("turtle_name_prefix", "turtle")

        self.spawn_frequency_ = self.get_parameter("spawn_frequency").value
        self.turtle_name_prefix_ = str(self.get_parameter(
            "turtle_name_prefix").value)
        self.use_sim_time_ = self.get_parameter("use_sim_time").value

        #Publishers, subscribers and services
        self.service_server_kill_turtle_ = self.create_service(
            KillTurtle, "/catch_turtle", self.turtle_killer)
        self.publisher_ = self.create_publisher(Turtles, "/alive_turtles", 10)
        self.publisher_timer = self.create_timer(1, self.callback_publisher)
        self.spawn_timer_ = self.create_timer(
            1/self.spawn_frequency_, self.turtle_spawner)

        self.turtle_number_ = 2

        # Array of Turtles objects
        self.turtles_alive_ = []

        self.turtle_spawner()
        self.get_logger().info("turtleSpawner is up!")

    # --------------------- Spawning methods ---------------------
    def turtle_spawner(self):

        self.spawn_service_client_ = self.create_client(Spawn, "/spawn")
        while not self.spawn_service_client_.wait_for_service(1.0):
            self.get_logger().warn("Waiting for Server /spawn...")

        x_ = round(random()*(10-1)+1, 2)
        y_ = round(random()*(10-1)+1, 2)
        theta_ = round(random()*2*pi, 2)
        name_ = self.turtle_name_prefix_ + str(self.turtle_number_)

        self.turtle_number_ += 1

        msg = Spawn.Request()
        '''
            float32 x
            float32 y
            float32 theta
            string name # Optional.  A unique name will be created and returned if this is empty
            ---
            string name
        '''

        msg.x = x_
        msg.y = y_
        msg.theta = theta_
        msg.name = name_

        future = self.spawn_service_client_.call_async(msg)

        future.add_done_callback(
            partial(self.callback_spawner, x=x_, y=y_, theta=theta_))

    def callback_spawner(self, future, x, y, theta):
        # Appending new turtle to turtles_alive_
        try:
            if future.result().name != "":
                new_turtle = Turtle()
                new_turtle.name = future.result().name
                new_turtle.x = x
                new_turtle.y = y
                new_turtle.theta = theta

                self.turtles_alive_.append(new_turtle)
            else:
                # if name is repeated -> name = ""
                self.get_logger().error("Turtle not created. Name repeated!")
        except:
            self.get_logger().error("Turtle not created. Error calling Spawn service!")



    # --------------------- Killing methods ---------------------
    def turtle_killer(self, request, response):
        #
        self.killer_service_client_ = self.create_client(Kill, "/kill")

        while not self.killer_service_client_.wait_for_service(1.0):
            self.get_logger().warn("Waiting for /kill service...")

        '''
        string name
        ---
        '''

        turtle_name = request.turtle

        if turtle_name != "":

            for turtle_robot in self.turtles_alive_:
                if turtle_robot.name == turtle_name:

                    msg = Kill.Request()
                    msg.name = turtle_name

                    self.get_logger().warn("Turtle to kill: " + str(turtle_name))

                    future = self.killer_service_client_.call_async(msg)

                    future.add_done_callback(
                        partial(self.callback_killer, turtleToKill=turtle_name))

                    response.success = True

                    break

            else:

                self.get_logger().warn("Turtle to kill already killed: " + str(turtle_name))
                response.success = False

        else:
            self.get_logger().error("Error")
            response.success = False

        return response

    def callback_killer(self, future, turtleToKill):
        self.get_logger().info(turtleToKill + " was killed!")

        # Remove turtle from turtles_alive_ list
        self.remove_turtle_from_list(turtleToKill)

        # Publish actualized turtles_alive_
        self.callback_publisher()

    def get_a_turtle(self):

        if self.turtles_alive_:
            turtle = self.turtles_alive_[0]
            if turtle.name:
                return turtle
            else:
                self.get_logger().error("There are no turtles alive!!!")
                return ""
        else:
            return ""

    def remove_turtle_from_list(self, turtle):
        for turtle_list in self.turtles_alive_:
            if turtle_list.name == turtle:
                self.turtles_alive_.remove(turtle_list)
                break

    # --------------------- Publisher methods ---------------------

    def callback_publisher(self):
        msg = Turtles()
        msg.turtles = self.turtles_alive_
        self.publisher_.publish(msg)


def main(args=None):
    rclpy.init(args=args)
    node = turtleSpawnerNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: turtlesim_catch_all_turtle_cpp/src/turtle_spawner.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/srv/spawn.hpp"
#include "turtlesim/srv/kill.hpp"
#include "turtlesim/msg/pose.hpp"
#include "my_robot_interfaces/srv/kill_turtle.hpp"
#include "my_robot_interfaces/msg/turtle.h"
#include "my_robot_interfaces/msg/turtles.hpp"
#include <ctime>

/*

    Functionality:

    Call the /spawn service to create a new turtle (choose random coordinates between 0.0 and 11.0 for both x and y),
    and call the /kill service to remove a turtle from the screen. Both those services are already advertised by the turtlesim_node.

    Publish the list of currently alive turtles with coordinates on a topic /alive_turtles.

    Handle a service server to “catch” a turtle, which means to call the /kill service and remove the turtle from the array of alive turtles.

*/

class turtleSpawnerNode : public rclcpp::Node
{
public:
    turtleSpawnerNode() : Node("turtle_spawner")
    {

        timer_ = this->create_wall_timer(std::chrono::seconds(10), std::bind(&turtleSpawnerNode::spawn_new_turtle, this));
        timer_publisher_ = this-> create_wall_timer(std::chrono::seconds(1),std::bind(&turtleSpawnerNode::publish_alive_turtles,this));
        server_ = this->create_service<my_robot_interfaces::srv::KillTurtle>("/catch_turtle", std::bind(&turtleSpawnerNode::callback_service_killer,
                                                                                                        this,
                                                                                                        std::placeholders::_1,
                                                                                                        std::placeholders::_2));

        client_kill_service_ = this->create_client<turtlesim::srv::Kill>("/kill");

        // Create publisher. Publish in topic /alive_turtles
        //  publish a Turtle() array.
        publisher_alive_turtles_ = this->create_publisher<my_robot_interfaces::msg::Turtles>("/alive_turtles",10);

        RCLCPP_INFO(this->get_logger(), "turtle_spawner is up!");
    }

    void spawn_new_turtle()
    {
        spawn_turtle_threads_.push_back(std::make_shared<std::thread>(std::bind(&turtleSpawnerNode::spawn_service, this)));
    }

    void spawn_service()
    {

        RCLCPP_INFO(this->get_logger(), "Executing spawn service");

        auto client = this->create_client<turtlesim::srv::Spawn>("spawn");

        float posX = random_number(0.0, 10.0);
        float posY = random_number(0.0, 10.0);

        // wait for service
        while (!client->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for server");
        }

        /*
            float32 x
            float32 y
            float32 theta
            string name # Optional.  A unique name will be created and returned if this is empty
            ---
            string name
        */

        // generate msg
        auto request = std::make_shared<turtlesim::srv::Spawn::Request>();

        request->x = posX;
        request->y = posY;
        request->name = prefix_turtle + std::to_string(counter_turtle);
        counter_turtle++;

        auto future = client->async_send_request(request);

        try
        {
            auto response = future.get();

            if (response->name != "")
            {
                // c_str allows me to use a char[] from a string
                RCLCPP_INFO(this->get_logger(), "Turtle was created!: %s ", response->name.c_str());

                /*
                    string name
                    float64 x
                    float64 y
                    float64 theta
                */

                // Add new turtle to the alive_turtles_vector_
                my_robot_interfaces::msg::Turtle new_turtle = my_robot_interfaces::msg::Turtle();

                new_turtle.name = response->name;
                new_turtle.x = request->x;
                new_turtle.y = request->y;
                new_turtle.theta = 0.0;
                
                alive_turtles_vector_.push_back(new_turtle);
                
            }
            else
            {
                RCLCPP_WARN(this->get_logger(), "Turtle was NOT created!: %s ", response->name.c_str());
            }
        }
        catch (std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Error in server Spawn!");
        }
    }

    float random_number(float min, float max)
    {
        float random_number = 0.0;

        random_number = (rand() / (double)RAND_MAX) * (max - min) + min;

        RCLCPP_INFO(this->get_logger(), "random_number: %f", random_number);

        return random_number;
    }

    void callback_service_killer(my_robot_interfaces::srv::KillTurtle::Request::SharedPtr request,
                                 my_robot_interfaces::srv::KillTurtle::Response::SharedPtr response)
    {
        /*
            string turtle
            ---
            bool success
        */
        std::string turtle_name = request->turtle;

        kill_turtle_threads_.push_back(std::make_shared<std::thread>(std::bind(&turtleSpawnerNode::kill_turtle, this, request->turtle)));

        response->success = true;
    }

    void kill_turtle(std::string request_name)
    {

        while (!client_kill_service_->wait_for_service(std::chrono::seconds(1)))
        {
            RCLCPP_WARN(this->get_logger(), "Waiting for server...");
        }

        auto request_turtlesim = std::make_shared<turtlesim::srv::Kill::Request>();

        /*
        string name
        ---
        */

        request_turtlesim->name = request_name;

        auto future = client_kill_service_->async_send_request(request_turtlesim);

        try
        {
            auto response_ = future.get();
            RCLCPP_INFO(this->get_logger(), "Turtle catched!");
            for (int i = 0; i < (int) alive_turtles_vector_.size() ;i++) 
            {
                if (alive_turtles_vector_.at(i).name == request_name)
                {
                    alive_turtles_vector_.erase(alive_turtles_vector_.begin() + i);
                    publish_alive_turtles();
                }
                break;
            }
        }
        catch (const std::exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "Error in service");
        }
    }

    void publish_alive_turtles()
    {
        auto msg = my_robot_interfaces::msg::Turtles();

        msg.turtles = alive_turtles_vector_;
        publisher_alive_turtles_->publish(msg);

    }

private:
    int counter_turtle = 2;
    std::string prefix_turtle = "turtle";
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::TimerBase::SharedPtr timer_publisher_;
    std::vector<std::shared_ptr<std::thread>> spawn_turtle_threads_;
    std::vector<std::shared_ptr<std::thread>> kill_turtle_threads_;
    std::vector<my_robot_interfaces::msg::Turtle> alive_turtles_vector_;

    rclcpp::Client<turtlesim::srv::Kill>::SharedPtr client_kill_service_;
    rclcpp::Service<my_robot_interfaces::srv::KillTurtle>::SharedPtr server_;

    rclcpp::Publisher<my_robot_interfaces::msg::Turtles>::SharedPtr publisher_alive_turtles_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<turtleSpawnerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
